README.md
# [Wikipedia Dark Theme](https://greasyfork.org/en/scripts/382833-wikipedia-dark-theme)

[![Codacy Badge](https://app.codacy.com/project/badge/Grade/960da3a5c8a44ccb9e35fcca45e5a88b)](https://www.codacy.com/gh/li-shangru/WikipediaDarkTheme/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=li-shangru/WikipediaDarkTheme&amp;utm_campaign=Badge_Grade)
[![CodeQL](https://github.com/li-shangru/WikipediaDarkTheme/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/li-shangru/WikipediaDarkTheme/actions/workflows/codeql-analysis.yml)
![GitHub last commit](https://img.shields.io/github/last-commit/li-shangru/WikipediaDarkTheme)
![GitHub](https://img.shields.io/github/license/li-shangru/WikipediaDarkTheme)

## Introduction

Wikipedia Dark Theme is a script to give Wikipedia pages a dark color theme.
This script inverts the text color and set the page background to black.
Providing a premium reading experience during both day and night.
Ease on the eyes, pleasant in the heart.

_Note: using this script WILL affect page loading speed; currently only optimized for English Wiki pages._

For a list of changes, please refer to [CHANGELOG.md](https://github.com/li-shangru/WikipediaDarkTheme/blob/master/CHANGELOG.md)

**Also available on : [Greasy Fork](https://greasyfork.org/en/scripts/382833-wikipedia-dark-theme)**

## Installation

Requires **[Tampermonkey](https://www.tampermonkey.net/)** or other user script manager.

To install:

* Click this [`Wikipedia Dark Theme.user.js`](https://github.com/li-shangru/WikipediaDarkTheme/raw/master/Wikipedia%20Dark%20Theme.user.js) file.

* An install page should pop up automatically.

## Settings

To open the Settings modal:

* Click the `Settings` option in the top right corner while browsing any Wiki pages.

Within, you can choose how Wikipedia pages look to you.

Select a single light or dark theme, or sync with your system and automatically switch between light and dark themes.

Moreover, you can choose your own link color.

Try to adjust these values and find the Wikipedia theme that looks best for you.

If you change your mind, you can always go back to default settings by clicking the `Restore defaults` button.

## Supported Pages

Wikipedia Dark Theme supports the following pages:

* [Wikipedia](https://en.wikipedia.org/wiki/Main_Page)
* [MediaWiki](https://www.mediawiki.org/wiki/MediaWiki)
* [Meta-Wiki](https://meta.wikimedia.org/wiki/Main_Page)
* [Wikispecies](https://species.wikimedia.org/wiki/Main_Page)
* [Wikibooks](https://en.wikibooks.org/wiki/Main_Page)
* [Wikidata](https://www.wikidata.org/wiki/Wikidata:Main_Page)
* [Wikimania](https://wikimania.wikimedia.org/wiki/Wikimania)
* [Wikinews](https://en.wikinews.org/wiki/Main_Page)
* [Wikiquote](https://en.wikiquote.org/wiki/Main_Page)
* [Wikisource](https://en.wikisource.org/wiki/Main_Page)
* [Wikiversity](https://en.wikiversity.org/wiki/Wikiversity:Main_Page)
* [Wikivoyage](https://en.wikivoyage.org/wiki/Main_Page)
* [Wiktionary](https://en.wiktionary.org/wiki/Wiktionary:Main_Page)

## Screenshots

| **Light Theme**                                          | **Dark Theme**                                                                   |
| :------------------------------------------------------: | :-------------------------------------------------------------------------: |
| ![A Wikipedia page](/screenshots/dark_theme_disabled.png)| ![A dark Wikipedia page](/screenshots/dark_theme_enabled.png)               |
| **Settings Menu**                                        | **Custom Link Color**                                                       |
| ![Settings menu](/screenshots/settings_menu.png)         | ![Wikipedia page with custom link color](/screenshots/custom_link_color.png)|


CHANGELOG.md
# Wikipedia Dark Theme Changelog

## [1.63](https://github.com/li-shangru/WikipediaDarkTheme/pull/15) (2022-08-26)

* Fixed issue https://github.com/li-shangru/WikipediaDarkTheme/issues/11
* Fixed issue https://github.com/li-shangru/WikipediaDarkTheme/issues/12
* Added a Monokai theme for Etherpad
* Thank you @csavard-sudo for contributing!

## [1.62](https://github.com/li-shangru/WikipediaDarkTheme/commit/9663fdf65a82d7f92a0aeadc37fdbaa103d6da09) (2022-08-23)

* Fixed issue https://github.com/li-shangru/WikipediaDarkTheme/issues/9
* Improved error handling

## [1.61](https://github.com/li-shangru/WikipediaDarkTheme/commit/2c2d15b5af1022c355054fc3edb15811238b0a01) (2022-02-08)

* Find an issue where settings button does not open settings modal in Vector 2022 skin

## [1.60](https://github.com/li-shangru/WikipediaDarkTheme/commit/3b4f643ce72a787874821225b54678eb5f8eacaf) (2022-02-02)

* Optimized the color display the following skins:
  * Vector
  * MinervaNeue
  * MonoBook
  * Timeless
  * Vector 2022
* Fixed a bug where settings option does not appear on the Vector 2022 skin

## [1.52](https://github.com/li-shangru/WikipediaDarkTheme/commit/8b257369dece32954804e7b07947b6a151b3b6a0) (2021-10-30)

* Background color improvement
* Changed default background color and relative luminance

## [1.51](https://github.com/li-shangru/WikipediaDarkTheme/commit/5f69dd7ebd34f9618879c684ab62c9f131cfe695) (2021-07-14)

* Fix list style
* Minor color change improvement

## [1.50](https://github.com/li-shangru/WikipediaDarkTheme/commit/b48f388b9a0bff9a3aa52ace93a7c0adeffcf1cc) (2021-07-11)

* Re-implemented color inverse function, from using RGB to match contrast value, to HSL matching relative luminance
* Visited links can now have different color than normal links
* Added custom color option for visited links
* Remove option to change contrast value
* Improvements on ancestries and table styles
* Source code reformat and structure improvement
* Fav icon fix

## [1.41](https://github.com/li-shangru/WikipediaDarkTheme/commit/8b60de0bc8cd7e3a5bcc5cee9bfadf35f9843bb3) (2021-07-04)

* Fixed a bug where when user dismissed settings modal, settings not reverted

## [1.40](https://github.com/li-shangru/WikipediaDarkTheme/commit/b55d757c33d49420d71e21d84eaf8e1e5315521c) (2021-07-04)

* Due to overwhelming (2) requests, I have added a settings modal. Within, user can change the theme and color preferences
* Minor bug fix

## [1.32](https://github.com/li-shangru/WikipediaDarkTheme/commit/fbd8c29372cbd3f53e0561b0d366b28ec9c18194) (2021-05-24)

* Hot fix

## [1.31](https://github.com/li-shangru/WikipediaDarkTheme/commit/04d152d0fbc8a53b93da45a1d0952af1067eba7d) (2021-05-24)

* Added more items to `EXCLUDE_SRC_TAG`

## [1.30](https://github.com/li-shangru/WikipediaDarkTheme/commit/31a4d593c449ec7f76fac343b98a5f7f5016e1c3) (2021-03-29)

* Minor performance improvement

## [1.29](https://github.com/li-shangru/WikipediaDarkTheme/commit/cd20efeacc4afdc7e80dce3c87d86ed826b32e53) (2021-03-24)

* Improvements on legend detection

## [1.28](https://github.com/li-shangru/WikipediaDarkTheme/commit/030b4a40fbe04e75f08d2539c6d438fad7054609) (2021-03-09)

* Added up/down arrow detection
* Bug fix on legend color
* Added more items to `EXCLUDE_SRC_TAG`

## [1.27](https://github.com/li-shangru/WikipediaDarkTheme/commit/febec449134b1e06dec1e842424fb394710b2ff4) (2021-03-08)

* Optimized family trees appearances, [example page](https://en.wikipedia.org/wiki/Family_tree_of_English_monarchs)

## [1.26](https://github.com/li-shangru/WikipediaDarkTheme/commit/297cc9afffcfce6a66e676831a73db9df51ce789) (2021-03-07)

* Minor function renaming
* Added some documentation

## [1.25](https://github.com/li-shangru/WikipediaDarkTheme/commit/ab1952f240969b1d06bfd05dadd73b5558901ca6) (2021-02-18)

* Fixed appearance for Español main page

## [1.24](https://github.com/li-shangru/WikipediaDarkTheme/commit/bbd3373e21125b3b229da5e8785c274408349f45) (2021-02-17)

* Added French translation for toggle script button
* Optimized the main pages of the following languages:
  * العربية
  * Deutsch
  * Español
  * Français
  * Italiano
  * Nederlands
  * 日本語
  * Polski
  * Português
  * Русский
  * Svenska
  * Українська
  * Tiếng Việt
  * 中文
  * 한국어
* Added more items to `EXCLUDE_SRC_TAG` and `INVERT_SRC_TAG`
* Updated screenshots

## [1.23](https://github.com/li-shangru/WikipediaDarkTheme/commit/47179528bf59ceb1cf1a2232baa95ed7e90fc4be) (2021-02-15)

* Small bug fix on locale detection

## [1.22](https://github.com/li-shangru/WikipediaDarkTheme/commit/f488ca4cbd9b5c82f1407d9bd690ac1c4eb16bfe) (2021-02-13)

* Wikipedia Dark Theme now supports the following additional pages:
  * MediaWiki
  * Meta-Wiki
  * Wikispecies
  * Wikibooks
  * Wikidata
  * Wikimania
  * Wikinews
  * Wikiquote
  * Wikisource
  * Wikiversity
  * Wikivoyage
  * Wiktionary
* Variable renaming and comments improvement
* Added more items to `EXCLUDE_SRC_TAG`

## [1.21](https://github.com/li-shangru/WikipediaDarkTheme/commit/d650db119f6f9c38c0936125656f56b070d4b6bc) (2021-02-13)

* Hotfix on `exclude_src_tag` items for sports team jersey display

## [1.20](https://github.com/li-shangru/WikipediaDarkTheme/commit/7dc8f9621060710e04a69b8c20ec070c851106b3) (2021-02-12)

* Fixed a bug where sports team jersey color not displaying correctly
* Optimized text background display
* General performance improvement

## [1.15](https://github.com/li-shangru/WikipediaDarkTheme/commit/c72dd157cad50616edcbe90858704f4c9a877ecb) (2020-12-28)

* Minor improvement

## [1.14](https://github.com/li-shangru/WikipediaDarkTheme/commit/cce68e486c8cca5610a77ac01906f6c0038f95ba) (2020-10-16)

* Minor function optimization

## [1.13](https://github.com/li-shangru/WikipediaDarkTheme/commit/bbcc7211daaf98da536599ec2257e33083b6ca55) (2020-08-31)

* Improved pie chart template detection

## [1.12](https://github.com/li-shangru/WikipediaDarkTheme/commit/3861bffb4b2fc925d79b374e0f743f386c8ed690) (2020-08-31)


Wikipedia Dark Mode.js
// ==UserScript==
// @name         Wikipedia Dark Mode
// @namespace    https://github.com/MaxsLi/WikiDarkMode/
// @version      0.6
// @icon         https://www.wikipedia.org/favicon.ico
// @description  Pure Dark theme for wikipedia.org
// @author       Shangru Li
// @match        *://*.wikipedia.org/*
// @grant        none
// ==/UserScript==

(function () {
    'use strict';
    // General idea is to put all elements on a wikipedia page to an array `allElements`
    // traverse through this array and reverse the color of each element accordingly
    // running time o(n), where n is the number of elements on a page <- improve needed
    var allElements = document.getElementsByTagName('*');

    for (var i = 0; i < allElements.length; i++) {
        var currentElement = allElements[i];
        // check for images
        if (currentElement.nodeName.toLowerCase() === 'img') {
            // Set images background color to white for better visibility
            currentElement.style.background = "rgb(255, 255, 255)";
            // skip to check next element
            continue;
        }
        // get the foreground color of the `currentElement`, using `getComputedStyle` will return thea actual showing
        // color of the given element.
        var foregroundColor = window.getComputedStyle(currentElement, null).getPropertyValue("color");
        var backgroundColor = window.getComputedStyle(currentElement, null).getPropertyValue("background-color");
        var r, g, b;
        var default_contrastValue = 8;
        var default_foregroundColor = "rgb(238, 255, 255)";
        var default_backgroundColor = "rgb(35, 35, 35)";
        // split the `default_backgroundColor` string to array of RGB values.
        var default_backgroundColor_array = default_backgroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/);

        // check if the `foregroundColor` is of type RGB value.
        if (foregroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/)) {
            // split color to RGB values.
            foregroundColor = foregroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/);
            // inverse color
            r = 255 - foregroundColor[1];
            g = 255 - foregroundColor[2];
            b = 255 - foregroundColor[3];
            // checking contrast between foregroundColor of `currentElement` and the `default_backgroundColor`
            // make sure the contrast is big enough to ensure a decent viewing experience.
            while (contrast([r, g, b], [default_backgroundColor_array[1], default_backgroundColor_array[2], default_backgroundColor_array[3]]) < default_contrastValue) {
                // increase each value by 30 each loop, i.e., increase the brightness of the current color
                r = r + 30;
                g = g + 30;
                b = b + 30;
            }
            // set color
            currentElement.style.color = 'rgb(' + r + ', ' + g + ', ' + b + ')';
        } else {
            // set default color
            currentElement.style.color = default_foregroundColor;
        }

        // check if the `backgroundColor` is of type RGB value.
        if (backgroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/)) {
            // split color to RGB values.
            backgroundColor = backgroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/);
            // inverse color
            r = 255 - backgroundColor[1];
            g = 255 - backgroundColor[2];
            b = 255 - backgroundColor[3];
            // for the background color one would not check for contrast, instead we increase the brightness of `backgroundColor`
            // if it is too dark, providing a comforting reading experience.
            if (r < default_backgroundColor_array[1] - 10 && g < default_backgroundColor_array[2] - 10 && g < default_backgroundColor_array[3] - 10) {
                r = r + 30;
                g = g + 30;
                b = b + 30;
            }
            // set color
            currentElement.style.backgroundColor = 'rgb(' + r + ', ' + g + ', ' + b + ')';
        } else {
            // set default color
            currentElement.style.backgroundColor = default_backgroundColor;
        }
    }

    // Helper function to calculate the contrast of two given color
    function luminance(r, g, b) {
        var a = [r, g, b].map(function (v) {
            v /= 255;
            return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
        });
        return a[0] * 0.2126 + a[1] * 0.7152 + a[2] * 0.0722;
    }

    function contrast(rgb1, rgb2) {
        return (luminance(rgb1[0], rgb1[1], rgb1[2]) + 0.05) / (luminance(rgb2[0], rgb2[1], rgb2[2]) + 0.05);
    }
})();


Wikipedia Dark Mode.user.js
// ==UserScript==
// @name         Wikipedia Dark Mode
// @namespace    https://github.com/MaxsLi/WikiDarkMode
// @version      0.61
// @icon         https://www.wikipedia.org/favicon.ico
// @description  Pure Dark theme for wikipedia.org
// @author       Shangru Li
// @match        *://*.wikipedia.org/*
// @grant        none
// ==/UserScript==

(function () {
    'use strict';
    // General idea is to put all elements on a wikipedia page to an array `allElements`
    // traverse through this array and reverse the color of each element accordingly
    // running time o(n), where n is the number of elements on a page <- improve needed
    var allElements = document.getElementsByTagName('*');

    for (var i = 0; i < allElements.length; i++) {
        var currentElement = allElements[i];
        // check for images
        if (currentElement.nodeName.toLowerCase() === 'img') {
            // Set images background color to white for better visibility
            currentElement.style.background = "rgb(255, 255, 255)";
            // skip to check next element
            continue;
        }
        // exception handler
        try {
            // get the foreground color of the `currentElement`, using `getComputedStyle` will return thea actual showing
            // color of the given element.
            var foregroundColor = window.getComputedStyle(currentElement, null).getPropertyValue("color");
            var backgroundColor = window.getComputedStyle(currentElement, null).getPropertyValue("background-color");
            var r, g, b;
            var default_contrastValue = 8;
            var default_foregroundColor = "rgb(238, 255, 255)";
            var default_backgroundColor = "rgb(35, 35, 35)";
            // split the `default_backgroundColor` string to array of RGB values.
            var default_backgroundColor_array = default_backgroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/);

            // check if the `foregroundColor` is of type RGB value.
            if (foregroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/)) {
                // split color to RGB values.
                foregroundColor = foregroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/);
                // inverse color
                r = 255 - foregroundColor[1];
                g = 255 - foregroundColor[2];
                b = 255 - foregroundColor[3];
                // checking contrast between foregroundColor of `currentElement` and the `default_backgroundColor`
                // make sure the contrast is big enough to ensure a decent viewing experience.
                while (contrast([r, g, b], [default_backgroundColor_array[1], default_backgroundColor_array[2], default_backgroundColor_array[3]]) < default_contrastValue) {
                    // increase each value by 30 each loop, i.e., increase the brightness of the current color
                    r = r + 30;
                    g = g + 30;
                    b = b + 30;
                }
                // set color
                currentElement.style.color = 'rgb(' + r + ', ' + g + ', ' + b + ')';
            } else {
                // set default color
                currentElement.style.color = default_foregroundColor;
            }

            // check if the `backgroundColor` is of type RGB value.
            if (backgroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/)) {
                // split color to RGB values.
                backgroundColor = backgroundColor.match(/rgb\((\d{1,3}), (\d{1,3}), (\d{1,3})\)/);
                // inverse color
                r = 255 - backgroundColor[1];
                g = 255 - backgroundColor[2];
                b = 255 - backgroundColor[3];
                // for the background color one would not check for contrast, instead we increase the brightness of `backgroundColor`
                // if it is too dark, providing a comforting reading experience.
                if (r < default_backgroundColor_array[1] - 10 && g < default_backgroundColor_array[2] - 10 && g < default_backgroundColor_array[3] - 10) {
                    r = r + 30;
                    g = g + 30;
                    b = b + 30;
                }
                // set color
                currentElement.style.backgroundColor = 'rgb(' + r + ', ' + g + ', ' + b + ')';
            } else {
                // set default color
                currentElement.style.backgroundColor = default_backgroundColor;
            }
        } catch (e){
            console.log(e)
        }
    }

    // Helper function to calculate the contrast of two given color
    function luminance(r, g, b) {
        var a = [r, g, b].map(function (v) {
            v /= 255;
            return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
        });
        return a[0] * 0.2126 + a[1] * 0.7152 + a[2] * 0.0722;
    }

    function contrast(rgb1, rgb2) {
        return (luminance(rgb1[0], rgb1[1], rgb1[2]) + 0.05) / (luminance(rgb2[0], rgb2[1], rgb2[2]) + 0.05);
    }
})();
